### PR TITLE
Add brainpool curves for EC key generation/cypher (main branch)

### PR DESCRIPTION
**Pull request by yklymenko**
The protocol for the communication between energy market participants in Germany enforce to use Brainpool P256r1 elliptic curve for signing and encryption of the soap parts (headers and attachments) .

This is a part (BDEW-profile) of the bigger AS4 Standart for the communication.

It would be nice to introduce this curves in santuario-xml-security to be able to use it with wss4j library.

The same as https://github.com/apache/santuario-xml-security-java/pull/292, but for main branch

### CODE REVIEW

**Comment by coheigea** [comment]
Looks fine to me. @seanjmullan can you review please?

**Review by coheigea — APPROVED** [review]

**Comment by seanjmullan** [comment]
Looks fine. It would be nice if there was also a test using the `javax.xml.crypto.dsig` API.

**Comment by yklymenko** [comment]
> Looks fine. It would be nice if there was also a test using the `javax.xml.crypto.dsig` API.

@seanjmullan you mean, something like org.apache.xml.security.test.dom.signature.EDDSASignatureTest ? I could try in the next days create such test

**Comment by seanjmullan** [comment]
> > Looks fine. It would be nice if there was also a test using the `javax.xml.crypto.dsig` API.
> 
> @seanjmullan you mean, something like org.apache.xml.security.test.dom.signature.EDDSASignatureTest ? I could try in the next days create such test

Yes, something like that would be great.

**Comment by yklymenko** [comment]
> > > Looks fine. It would be nice if there was also a test using the `javax.xml.crypto.dsig` API.
> > 
> > 
> > @seanjmullan you mean, something like org.apache.xml.security.test.dom.signature.EDDSASignatureTest ? I could try in the next days create such test
> 
> Yes, something like that would be great.

@seanjmullan  I've added the test.  I have some problem to run org.apache.xml.security.utils.KeyUtilsTest unter jdk17 with profile bouncecastle. The "DH" Case for generateEphemeralDHKeyPair  seems to be broken. org.apache.xml.security.utils.KeyUtils.KeyType#DH declaration use 1.2.840.113549.1.3.1 oid, but 1.2.840.10046.2.1 seems to be the correct one (http://oid-info.com/get/1.2.840.10046.2.1). Or may be I'm wrong?

**Comment by jrihtarsic** [comment]
@yklymenko  good catch: Looks like  SunJCE returns  different OID for generated (and then encoded) keys with algorithm name  'DH' than the BouncyCastle:

The result for test below is :

Expected :1.2.840.113549.1.3.1
Actual   :1.2.840.10046.2.1
   ```
@Test
    void generateEphemeralDHKeyPairComparison() throws Exception {
        // gen DH key pair with SunJCE
        KeyPairGenerator kpgSunJCE = KeyPairGenerator.getInstance("DH");
        KeyPair keyPairSun = kpgSunJCE.generateKeyPair();

        // gen DH key pair with BC
        KeyPairGenerator kpgBC = KeyPairGenerator.getInstance("DH", new org.bouncycastle.jce.provider.BouncyCastleProvider());
        KeyPair keyPairBC = kpgBC.generateKeyPair();

        String keyOidSunJCE = DERDecoderUtils.getAlgorithmIdFromPublicKey(keyPairSun.getPublic());
        String keyOidBC = DERDecoderUtils.getAlgorithmIdFromPublicKey(keyPairBC.getPublic());

        Assertions.assertEquals(keyOidSunJCE, keyOidBC);
    }
```

**Comment by yklymenko** [comment]
> My suggestion is To remove DH("DH", "PKCS #3", KeyAlgorithmType.DH, "1.2.840.113549.1.3.1"), From KeyType
> 
> The Diffie-Hellman key agreement using RSA keys is gradually becoming obsolete and currently, it is not supported by the xmlsec key agreement method implementation. I included it there primarily for the sake of completeness, anticipating that someone might (but not very likely) add support for Diffie-Hellman in the future.
> 
> Beside CodeQL marks it as potentially unsecure: image

In general, I've looked in BC how they decide, which oid should be used. They have both oid's as aliases and select one of them depending on configuration in org.bouncycastle.jcajce.provider.asymmetric.dh.BCDHPublicKey#getEncoded (see both return statements)
Sure, I can delete two or three lines to make it green again, but I'm not sure, that this should be a part of this PR

**Comment by jrihtarsic** [comment]
@yklymenko, you have a good point. 
To make it transparent I created now the SANTUARIO-613 and the PR #297

**Comment by seanjmullan** [comment]
> > > > Looks fine. It would be nice if there was also a test using the `javax.xml.crypto.dsig` API.
> > > 
> > > 
> > > @seanjmullan you mean, something like org.apache.xml.security.test.dom.signature.EDDSASignatureTest ? I could try in the next days create such test
> > 
> > 
> > Yes, something like that would be great.
> 
> @seanjmullan I've added the test. I have some problem to run org.apache.xml.security.utils.KeyUtilsTest unter jdk17 with profile bouncecastle. The "DH" Case for generateEphemeralDHKeyPair seems to be broken. org.apache.xml.security.utils.KeyUtils.KeyType#DH declaration use 1.2.840.113549.1.3.1 oid, but 1.2.840.10046.2.1 seems to be the correct one (http://oid-info.com/get/1.2.840.10046.2.1). Or may be I'm wrong?

This is still using the Santuario API (org.apache.xml.security). Sorry, I missed the package name of the test you said above when I said that would be great. 

What I meant is something like test/java/org/apache/xml/security/test/javax/xml/crypto/dsig/XMLSignatureEdDSATest.java

**Comment by jrihtarsic** [comment]
@yklymenko @seanjmullan 
I can do that. I have already internal task to do verify it for eDelivery AS4 profile and I can 
implement also the required tests to santuario. I created a task:
https://issues.apache.org/jira/browse/SANTUARIO-614 
And I will provide the PR today for it.

File: src/test/java/org/apache/xml/security/test/dom/encryption/XMLCipherBrainpoolTest.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        // prepare the test document
+        Document d = document(); // source
+        Document ed = null;
+        Document dd = null;
+        Element e = (Element) d.getElementsByTagName(element()).item(index());
+        Element ee = null;
+        String source = null;
+        String target = null;
+
+        source = toString(d);
```
Comment (seanjmullan) [inline src/test/java/org/apache/xml/security/test/dom/encryption/XMLCipherBrainpoolTest.java:166]: Can combine lines 163 and 166.

File: src/test/java/org/apache/xml/security/test/dom/encryption/XMLCipherBrainpoolTest.java [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            builder.setKeyInfo(builderKeyInfo);
+        }
+        builderKeyInfo.add(encryptedKey);
+
+        ed = cipherEncData.doFinal(d, e);
+
+        Files.write(Paths.get("target","test-enc-"+keyType.name()+".xml"), toString(ed).getBytes());
+
+        //decrypt
+        ee = (Element) ed.getElementsByTagName("xenc:EncryptedData").item(0);
```
Comment (seanjmullan) [inline src/test/java/org/apache/xml/security/test/dom/encryption/XMLCipherBrainpoolTest.java:213]: Can combine lines 162 and 213.

File: src/test/java/org/apache/xml/security/test/dom/encryption/XMLCipherBrainpoolTest.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        XMLCipher cipherDecData = XMLCipher.getInstance(dataEncryptionAlgorithm);
+        cipherDecData.init(XMLCipher.DECRYPT_MODE, null);
+        cipherDecData.setKEK(privRecipientKey);
+        cipherDecData.setSecureValidation(true);
+        dd = cipherDecData.doFinal(ed, ee);
```
Comment (seanjmullan) [inline src/test/java/org/apache/xml/security/test/dom/encryption/XMLCipherBrainpoolTest.java:218]: Can combine lines 160 and 218.

File: src/test/java/org/apache/xml/security/test/dom/encryption/XMLCipherBrainpoolTest.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        EncryptedData builder = cipherEncData.getEncryptedData();
+        // add encrypted key to key info in encrypted data
+        KeyInfo builderKeyInfo = builder.getKeyInfo();
+        if (builderKeyInfo == null) {
+            builderKeyInfo = new KeyInfo(d);
+            builder.setKeyInfo(builderKeyInfo);
+        }
+        builderKeyInfo.add(encryptedKey);
+
+        ed = cipherEncData.doFinal(d, e);
```
Comment (seanjmullan) [inline src/test/java/org/apache/xml/security/test/dom/encryption/XMLCipherBrainpoolTest.java:208]: Can combine lines 159 and 208.

File: src/test/java/org/apache/xml/security/test/dom/signature/ECBrainpoolSignatureTest.java [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+
+/**
+ * Tests that creates and verifies EC signatures with brainpool curves.
+ *
+ */
+class ECBrainpoolSignatureTest {
+
+    private static final String ECDSA_JKS =
+        "src/test/resources/org/apache/xml/security/samples/input/ecbrainpool.jks";
```
Comment (seanjmullan) [inline src/test/java/org/apache/xml/security/test/dom/signature/ECBrainpoolSignatureTest.java:65]: Can you create this KeyStore as part of the test? We should stop storing binary files in the repository. I know that there are already existing ones, but I would like to stop doing that going forward.

**Pull request merged**